src/compiler/crystal/codegen/primitives.cr

```diff
@@ -173,6 +174,90 @@ class Crystal::CodeGenVisitor
     end
   end
 
+  def codegen_binary_op_add(t : IntegerType, t1, t2, p1, p2)
+    # TODO remove on 0.28
+    return codegen_trunc_binary_op_result(t1, t2, builder.add(p1, p2)) unless @program.has_flag?("preview_overflow")
+
+    llvm_fun = case t.kind
+               when :i8
+                 binary_overflow_fun "llvm.sadd.with.overflow.i8", llvm_context.int8
+               when :i16
+                 binary_overflow_fun "llvm.sadd.with.overflow.i16", llvm_context.int16
+               when :i32
+                 binary_overflow_fun "llvm.sadd.with.overflow.i32", llvm_context.int32
+               when :i64
+                 binary_overflow_fun "llvm.sadd.with.overflow.i64", llvm_context.int64
+               when :i128
+                 binary_overflow_fun "llvm.sadd.with.overflow.i128", llvm_context.int128
+               when :u8
+                 binary_overflow_fun "llvm.uadd.with.overflow.i8", llvm_context.int8
+               when :u16
+                 binary_overflow_fun "llvm.uadd.with.overflow.i16", llvm_context.int16
+               when :u32
+                 binary_overflow_fun "llvm.uadd.with.overflow.i32", llvm_context.int32
+               when :u64
+                 binary_overflow_fun "llvm.uadd.with.overflow.i64", llvm_context.int64
+               when :u128
+                 binary_overflow_fun "llvm.uadd.with.overflow.i128", llvm_context.int128
+               else
+                 raise "unreachable"
+               end
```

Comment:
I suggest modifying `binary_overflow_fun` to take `"add"` and the `IntegerType` and work out the intrinsic name and type. Then, you can make a `codegen_binary_op_result("add", tmax, t1, t2, p1, p2)`, which additionally uses the `"add"` to work out which `codegen_trunc_binary_op_result` to use when overflow checking is disabled.

Reply:
I like to have the intrinsics explicitly written so the can be searched. I looked for explicitly avoiding the concatenation. Other alternative would be to leave some of the names of the intrinsics commented in the function that will return them.